rfc/closeable_tracer.md

```diff
@@ -0,0 +1,37 @@
+# Closeable Tracers
+
+**Current State:** Draft  
+**Author:** [austinlparker](https://github.com/austinlparker)
+
+This proposal builds off of discussions primarily in the Java API surrounding the inclusion of a **Close** method in that library. The primary goal of adding this to the specification is to provide a clean decoupling of OpenTracing `Tracer`s from their concrete implementations in a consistent way across platforms and languages.
+
+# Background
+Currently, a `Tracer` has three major responsibilities - creating `Span`s and performing de/serialization of them across boundaries via `Extract` and `Inject`. In practice, a concrete implementation of a `Tracer` has more responsibilities; for example, appending created `Span`s to a buffer, or flushing that buffer through a wire format to some listening system that records traces. One very common method implemented by a concrete `Tracer` is `Close`. The implementation, and utility, of this method varies by language. In general,
+calling `Close` on an instantiated `Tracer` will finalize traces and release any resources held by the `Tracer`. 
+
+Discussions (See [this thread](https://github.com/opentracing/opentracing-java/issues/250) for example) around the inclusion of a `Close` method in Java have crystallized around including this feature in the Java interface. Specifically, there would be value in having `Close` specified on the OpenTracing interface in order to prevent a cast to the concrete tracer, which cannot occur in a situation where the implementation is being loaded dynamically. This situation exists in C# as well, where a similar pattern of try-with-resource (called `using`) exists.
+
+# Specification Changes
+The `Tracer` interface is extended with the following:
+
+* **Close**, a method that is responsible for closing the active tracer. For `Tracer`s that record `Span`s for persistance, calling **Close** should flush the current in-memory collection to the configured persistance store. For stateless `Tracer`s or similar, this can be a NoOp. The **Close** method should be considered idempotent; closing an already closed `Tracer` should not raise an error. `Span`s that are created or finished after a `Tracer` has been closed may or may not be flushed.
```

Comment:
Can we add a comment like:

> The `Tracer` interface should extend the platform's idiomatic "close" interface, such as Java's [`Closeable`](https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html).

This would allow the tracer to be used with a try-with-resources construct, similar to the C# example you mention afterwards.

Reply:
I already indicate that Java/C# should use idomatic close in the 'use case' section, not sure if we want to have language-specific parts in the spec description?

Reply:
No, I should have read more carefully :-)